src/Analyzer/Resolve/QueryAnalyzer.cpp

```diff
@@ -2958,27 +2958,28 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
             /// Replace storage with values storage of insertion block
             if (StoragePtr storage = scope.context->getViewSource())
             {
-                QueryTreeNodePtr table_expression;
+                QueryTreeNodePtr table_expression = in_second_argument;
+
                 /// Process possibly nested sub-selects
-                for (auto * query_node = in_second_argument->as<QueryNode>(); query_node; query_node = table_expression->as<QueryNode>())
-                    table_expression = extractLeftTableExpression(query_node->getJoinTree());
+                while (table_expression)
+                {
+                    if (auto * query_node = table_expression->as<QueryNode>())
+                        table_expression = extractLeftTableExpression(query_node->getJoinTree());
+                    else if (auto * union_node = table_expression->as<UnionNode>())
+                        table_expression = union_node->getQueries().getNodes().at(0);
+                    else
+                        break;
+                }
 
-                if (table_expression)
+                auto * table_expression_table_node = table_expression->as<TableNode>();
```

Comment:
`table_expression` can be `nullptr` at this point. Let's move the check before cast.

Reply:
Updated.